### PR TITLE
util, types: don't let SPM be affected by charset (#23161)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #23161 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/23295
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/23295:release-4.0-2bea06ed238c
```

---

### What problem does this PR solve?

Issue Number: close #xxx 

Different default charsets in different drivers(or versions) make SQL bind cannot work.
So we don't let SPM be affected by charset.

We add a flag `RestoreStringWithoutCharset` to control this.

Need to merge [parser pr](https://github.com/pingcap/parser/pull/1184) first

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- don't let SPM be affected by charset

### CODE REVIEW

**Review by eurekaka — COMMENTED** [review]
@rebelice please resolve conflicts

File: session/bootstrap.go [hunk 1]
```diff
@@ -1148,7 +1148,7 @@ type bindInfo struct {
 }
 
 func upgradeToVer51(s Session, ver int64) {
-	if ver >= version51 {
+	if ver >= version52 {
```
Comment (Reminiscent) [inline session/bootstrap.go:1151]: Why not `version51` here? Rest LGTM

**Review by eurekaka — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]

**Review by lzmhhh123 — COMMENTED** [review]
A lot of test cases are failed.

**Pull request merged**